### PR TITLE
Add :default cases to multi-methods

### PR DESCRIPTION
**Pull request by arichiardi**
I had a bit more time and setup the `:default` for `url-dir` as well (and changed the commit msg).

### CODE REVIEW

**Comment by weavejester** [comment]
Thanks, but can you remove the changes to `url-dir`? That's a separate issue, I think. A predicate should always return a boolean value, but in other functions it's generally better to throw an exception on unexpected input.

Also can you change `closes` to `fixes` in the commit message, as I mentioned in #1?

**Comment by arichiardi** [comment]
Isn't `url-dir` a function returning things? I just thought that when the input is `nil`, the output should be `nil` as well. Sure for the comment.

**Comment by weavejester** [comment]
Right, but I think it's a different case to `directory?`. With `directory?`, a URL will either be a directory or it won't be. With `url-dir` I think we want to distinguish between a valid URL that contains no directories, and a URL that cannot be searched for directories.

File: test/resauce/core_test.clj [hunk 1]
```diff
@@ -16,6 +17,8 @@
     (is (re-find #"test/resauce$" (second rs)))))
 
 (deftest test-url-dir
+  (is (nil? (url-dir nil)))
+  (is (thrown? java.lang.IllegalArgumentException (url-dir (io/as-url "http://www.example.com/docs/resource1.html"))))
```
Comment (weavejester) [inline test/resauce/core_test.clj:21]: Can you split this line up so that it's not (much) over 80 characters?

Comment (arichiardi) [reply test/resauce/core_test.clj:21]: Sure

**Comment by arichiardi** [comment]
I agree with you, however I guess you agree that the base case for the implementation of the protocol on `nil` is not different from what I added. Unless there is something I am missing of course 😀

Anyway it is not a problem for me to split things, so that you can add more details on a patch that handles more cases in `url-dir`.

I will revive the other branch.



-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Pull request closed without merging**